### PR TITLE
array: add shift function

### PR DESCRIPTION
**Pull request by vchikalkin**
Request to add **shift** function:

Example: 
I have array: [1, 2, 3, 4, 5, 6, 7, 8, 9];
I want to shift array n positions right/left

Usage:

```
const arr = [1, 2, 3, 4, 5, 6, 7, 8, 9];

const shiftedArr = shift(arr, 3); // [7, 8, 9, 1, 2, 3, 4, 5, 6]
const shiftedLeftArr = shift(arr,-3); // [4, 5, 6, 7, 8, 9, 1, 2, 3]
```

### CODE REVIEW

**Review by matteosacchetto — COMMENTED** [review]
What if the number of steps is greater than the number of elements in the array?
What should happen?

Example:
```js
const arr = [1, 2, 3, 4, 5, 6, 7, 8, 9];
shift(arr, 15) // What is the expected behavior?
shift(arr, -15) // What is the expected behavior?
```

**Comment by vchikalkin** [comment]
> What if the number of steps is greater than the number of elements in the array? What should happen?
> 
> Example:
> 
> ```js
> const arr = [1, 2, 3, 4, 5, 6, 7, 8, 9];
> shift(arr, 15) // What is the expected behavior?
> shift(arr, -15) // What is the expected behavior?
> ```

nothing bad will happen, the function will just return the array in its original state:

const arr = [1, 2, 3, 4, 5, 6, 7, 8, 9];
shift(arr, 15) // [1, 2, 3, 4, 5, 6, 7, 8, 9]
shift(arr, -15) // [1, 2, 3, 4, 5, 6, 7, 8, 9]

but you're right. need to add check for shift number and targe array length

**Comment by matteosacchetto** [comment]
What I was thinking is the following, what if we allow using higher numbers e shift in a circular way?

**Comment by matteosacchetto** [comment]
Something like this:
```ts
export function shift<T>(arr: ReadonlyArray<T>, n: number) {
  if (n === 0) return arr;

  const nPos = (-n % arr.length + arr.length) % arr.length;
  return [...arr.slice(pos, arr.length), ...arr.slice(0, pos)];
}
```

**Comment by vchikalkin** [comment]
> What I was thinking is the following, what if we allow using higher numbers e shift in a circular way?

I think it should look something like this:
```
function shift<T>(arr: ReadonlyArray<T>, n: number): Array<T> {
  if (n === 0) return arr;

  const shiftNumber = n % arr.length;

  return [
    ...arr.slice(-shiftNumber, arr.length),
    ...arr.slice(0, -shiftNumber),
  ];
}
```

So, in your examples the real shift number is 6 and results will be:
const arr = [1, 2, 3, 4, 5, 6, 7, 8, 9];
shift(arr, 15)  // [ 4, 5, 6, 7, 8, 9, 1, 2, 3 ]
shift(arr, -15) //[ 7, 8, 9, 1, 2, 3, 4, 5, 6 ]

**Comment by matteosacchetto** [comment]
The idea is that, if you have 9 elements, and shift by 9, you obtain the same results as not shifting, while if you shift by 10, you are actually shifting by 1



> > What I was thinking is the following, what if we allow using higher numbers e shift in a circular way?
> 
> I think it should look something like this:
> 
> ```
> function shift<T>(arr: ReadonlyArray<T>, n: number): Array<T> {
>   if (n === 0) return arr;
> 
>   const shiftNumber = n % arr.length;
> 
>   return [
>     ...arr.slice(-shiftNumber, arr.length),
>     ...arr.slice(0, -shiftNumber),
>   ];
> }
> ```
> 
> So, in your examples the real shift number is 6 and results will be:
> 
> const arr = [1, 2, 3, 4, 5, 6, 7, 8, 9]; shift(arr, 15) _// [ 4, 5, 6, 7, 8, 9, 1, 2, 3 ]_ shift(arr, -15) _//[ 7, 8, 9, 1, 2, 3, 4, 5, 6 ]_

Yes, something like that seems good to me

**Comment by vchikalkin** [comment]
> The idea is that, if you have 9 elements, and shift by 9, you obtain the same results as not shifting, while if you shift by 10, you are actually shifting by 1

Exactly!
So, we can update the check with the following to avoid unnecessary calculations: 
`  if (n === 0 || arr.length === n) return arr;`

**Comment by matteosacchetto** [comment]
Maybe, that control is even better after evaluating this: `const shiftNumber = n % arr.length;`, so that if I pass a multiple of the array.length it won't do anything

**Comment by vchikalkin** [comment]
> It is missing the negative case where `n === -arr.length`

And `arr.length === 0`

**Comment by vchikalkin** [comment]
@matteosacchetto 
I uploaded the changes

File: src/array.ts [hunk 1]
```diff
@@ -396,3 +396,16 @@ export const diff = <T>(
...
+
+/**
+ * Shift array items by n steps
+ * If n > 0 items will shift n steps to the right
+ * If n < 0 items will shift n steps to the left
+ */
+ export function shift<T>(arr: Array<T>, n: number) {
+  if (n === 0 || arr.length === 0 || arr.length === n || n === -arr.length) return arr;
+
+  const shiftNumber = n % arr.length;
```
Comment (matteosacchetto) [inline src/array.ts:408]: Not totally convinced by it, since if I set n to a multiple of arr.length then it will do the computing operations
Here is a suggestion

```ts
  if (arr.length === 0) return arr;
  const shiftNumber = n % arr.length;
  if(shiftNumber === 0) return arr;
```

Comment (matteosacchetto) [reply src/array.ts:408]: This would match also multiples of `arr.length` (positive and negatice) as well as 0

**Review by matteosacchetto — CHANGES_REQUESTED** [review]

File: src/array.ts (hunk 1 above)
Comment (vchikalkin) [reply src/array.ts:408]: Then the implementation is even simpler and looks prettier:
```
  if (n === 0 || arr.length === 0) return arr;
  const shiftNumber = n % arr.length;
  if (shiftNumber === 0) return arr;
```

Comment (vchikalkin) [reply src/array.ts]: Agree and upload

Comment (matteosacchetto) [reply src/array.ts]: Great! Now, simply add a test where you pass an empty array and the pipeline should not fail 😉

**Comment by matteosacchetto** [comment]
It is a simple issue of a missing test.
Add a test for the empty array case and it should not complain

**Review by matteosacchetto — APPROVED** [review]

**Comment by vchikalkin** [comment]
@matteosacchetto 
Now it's ok! 😉

**Review by matteosacchetto — APPROVED** [review]

**Comment by matteosacchetto** [comment]
Great! 😉

**Review by rayepps — APPROVED** [review]
Awesome PR, thank you @vchikalkin !! And thank you @matteosacchetto for the teamwork! This will go out in the next published version 🎉
